### PR TITLE
zcash_address 0.3.2: ZIP 320 TEX address support

### LINKED ISSUE #1220: zcash_address: Implement ZIP 320

### LINKED ISSUE #1221: zcash_address 0.3.2

### PR DESCRIPTION
**Pull request by str4d**
Closes #1220.
Closes #1221.

### CODE REVIEW

**Review by nuttycom — APPROVED** [review]
utACK

File: components/zcash_address/src/convert.rs [hunk 1]
```diff
@@ -353,6 +373,17 @@ impl ToAddress for ZcashAddress {
...
         }
     }
+
+    fn from_tex(net: Network, data: [u8; 20]) -> Self {
+        ZcashAddress {
+            net: if let Network::Regtest = net {
+                Network::Test
+            } else {
+                net
+            },
```
Comment (str4d) [inline components/zcash_address/src/convert.rs:383]: I'm confused as to why ZIP 320 does not specify a regtest prefix, but I don't particularly care myself

Comment (nuttycom) [reply components/zcash_address/src/convert.rs:383]: Oh, you missed the commit that @daira added that includes regtest support.

**Review by nuttycom — CHANGES_REQUESTED** [review]
This also needs 4992ff765f68871343b1ad8c25363486e78a0264

**Comment by str4d** [comment]
[Force-pushed](https://github.com/zcash/librustzcash/compare/dd9c2d510f74b04b501854957b316e8571432a7d..fd060d8987abccae29bcb5bb3fa6f5f7a931ad7d) to include the regtest prefix from @daira's review on #1135.

File: components/zcash_address/src/convert.rs (hunk 1 above)
Comment (str4d) [reply components/zcash_address/src/convert.rs:383]: Done.

Comment (nuttycom) [reply components/zcash_address/src/convert.rs:383]: Hm, I think that Daira's PR also missed this spot.

**Review by nuttycom — CHANGES_REQUESTED** [review]
@str4d the location you originally flagged still needs to be fixed.

File: components/zcash_address/src/convert.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (str4d) [reply components/zcash_address/src/convert.rs:383]: Good point, this was not included in @daira's comments.

**Comment by str4d** [comment]
[Force-pushed](https://github.com/zcash/librustzcash/compare/fd060d8987abccae29bcb5bb3fa6f5f7a931ad7d..170a1018391bb445fe179236d1c562918d74c845) to fix the regtest issue in `impl ToAddress for ZcashAddress` now that we have regtest prefixes.

**Review by nuttycom — APPROVED** [review]
utACK

File: components/zcash_address/src/lib.rs [hunk 2]
```diff
@@ -270,6 +272,7 @@ impl ZcashAddress {
                 T::try_from_raw_transparent_p2pkh(data)
             }
             AddressKind::P2sh(data) if regtest_exception => T::try_from_raw_transparent_p2sh(data),
+            AddressKind::Tex(data) if regtest_exception => T::try_from_raw_tex(data),
```
Comment (str4d) [inline components/zcash_address/src/lib.rs:275]: Another spot that needs fixing after regtest prefixes were defined:
```suggestion
            AddressKind::Tex(data) if network_matches => T::try_from_raw_tex(data),
```

**Comment by str4d** [comment]
[Force-pushed](https://github.com/zcash/librustzcash/compare/170a1018391bb445fe179236d1c562918d74c845..2c5bc9ba7a7df6cb46ea4c2f3e226e31b7d2625a) to fix the other bug after a regtest prefix was defined.

**Review by nuttycom — APPROVED** [review]
re-utACK

**Review by daira — DISMISSED** [review]
utACK

**Comment by str4d** [comment]
[Force-pushed](https://github.com/zcash/librustzcash/compare/ba3dcd57ec363ac130b4ec5f6ae525ca218f8ccd..9758bee9df5f6f426061618e9b9ecf0b5083d853) to update the `zcash_protocol` changelog.

**Review by nuttycom — APPROVED** [review]
utACK 9758bee9df5f6f426061618e9b9ecf0b5083d853

**Comment by str4d** [comment]
[Force-pushed](https://github.com/zcash/librustzcash/compare/9758bee9df5f6f426061618e9b9ecf0b5083d853..d539f04f2e789552ab20b7a1e8a87a0d1005c394) to add a `zcash_protocol 0.1.0` release commit prior to the `zcash_address 0.3.2` release commit.

**Review by nuttycom — APPROVED** [review]
re-utACK

**Review by daira — COMMENTED** [review]
Post-hoc ACK

**Pull request merged**